File: day2/java/DLL.java
package day2.java;

import java.util.NoSuchElementException;

/*
 * Double Linked List
 */
public class DLL implements List {
    private int size;

    private class Node {
        public int value;
        public Node prev;
        public Node next;

        public Node(int value, Node prev, Node next) {
            this.value = value;
            this.prev = prev;
            this.next = next;
        }

        public Node() {
            this(0, null, null);
        }
    }

    // Critical to have a sentinel node
    // can set any value other than -1
    private Node sentinel = new Node(-1, null, null);

    public DLL() {
        sentinel.prev = sentinel;
        sentinel.next = sentinel;
        size = 0;
    }

    @Override
    public void add(int index, int value) {
        if (index < 0 || index > size) {
            throw new IndexOutOfBoundsException("Index out of bounds");
        }

        Node current = sentinel;
        for (int i = 0; i < index; i++) {
            current = current.next;
        }

        Node newNode = new Node(value, current, current.next);
        current.next.prev = newNode;
        current.next = newNode;
        size++;
    }

    @Override
    public void addFirst(int value) {
        add(0, value);
    }

    @Override
    public void addLast(int value) {
        // TODO: make this method as efficient as possible
        Node newNode = new Node(value, sentinel.prev, sentinel);
        sentinel.prev.next = newNode;
        sentinel.prev = newNode;
        size++;
    }

    @Override
    public int get(int index) {
        if (index < 0 || index >= size) {
            throw new IndexOutOfBoundsException("Index out of bounds");
        }

        Node current = sentinel.next;
        for (int i = 0; i < index; i++) {
            current = current.next;
        }

        return current.value;
    }

    @Override
    public int getFirst() {
        return get(0);
    }

    @Override
    public int getLast() {
        // TODO: This is not efficient, rewrite this and make it efficient
        if (isEmpty()) {
            throw new NoSuchElementException("List is empty");
        }
        return sentinel.prev.value;
    }

    @Override
    public int remove(int index) {
        if (index < 0 || index >= size) {
            throw new IndexOutOfBoundsException("Index out of bounds");
        }

        Node current = sentinel.next;
        for (int i = 0; i < index; i++) {
            current = current.next;
        }

        int removedValue = current.value;
        current.prev.next = current.next;
        current.next.prev = current.prev;
        size--;
        return removedValue;
    }

    @Override
    public int removeFirst() {
        return remove(0);
    }

    @Override
    public int removeLast() {
        // TODO: This is not efficient, rewrite this and make it efficient
        if (isEmpty()) {
            throw new NoSuchElementException("List is empty");
        }

        int removedValue = sentinel.prev.value;
        sentinel.prev = sentinel.prev.prev;
        sentinel.prev.next = sentinel;
        size--;
        return removedValue;
    }

    @Override
    public int size() {
        return this.size;
    }

    @Override
    public boolean isEmpty() {
        return this.size == 0;
    }

    @Override
    public int indexOf(int value) {
        // TODO: Implement this method

        Node current = sentinel.next;
        int index = 0;

        while (current != sentinel) {
            if (current.value == value) {
                return index;
            }
            current = current.next;
            index++;
        }

        return -1;
    }

    @Override
    public boolean contains(int value) {
        // TODO: Implement this method

        Node current = sentinel.next;

        while (current != sentinel) {
            if (current.value == value) {
                return true;
            }
            current = current.next;
        }

        return false;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append("[");
        Node current = sentinel.next;
        while (current != sentinel) {
            sb.append(current.value);
            if (current.next != sentinel) {
                sb.append(", ");
            }
            current = current.next;
        }
        sb.append("]");
        return sb.toString();
    }

    // Helper method to convert an array to a DLL
    public static DLL ArrayToDLL(int[] arr) {
        DLL dll = new DLL();
        for (int i = 0; i < arr.length; i++) {
            dll.addLast(arr[i]);
        }
        return dll;
    }

    public static void main(String[] args) {
        int[] arr = { 1, 2, 3, 4, 5 };
        DLL dll = ArrayToDLL(arr);
        System.out.println(dll);

        System.out.println(dll.size());
        System.out.println(dll.isEmpty());

        System.out.println(dll.indexOf(3));
        System.out.println(dll.contains(3));
        System.out.println(dll.indexOf(10));
        System.out.println(dll.contains(10));

        dll.add(2, 10);
        System.out.println(dll);

        dll.addFirst(20);
        System.out.println(dll);

        dll.addLast(30);
        System.out.println(dll);

        System.out.println(dll.get(2));
        System.out.println(dll.getFirst());
        System.out.println(dll.getLast());

        System.out.println(dll.remove(2));
        System.out.println(dll);
        System.out.println(dll.removeFirst());
        System.out.println(dll);
        System.out.println(dll.removeLast());
        System.out.println(dll);
    }
}

File: day2/javascript/dll.js
class Node {
  constructor(value = -1, prev = null, next = null) {
    this.value = value;
    this.prev = prev;
    this.next = next;
  }
}

class DLL {
  constructor() {
    this.sentinel = new Node(-1);
    this.sentinel.prev = this.sentinel;
    this.sentinel.next = this.sentinel;
    this.size = 0;
  }

  add(index, value) {
    if (index < 0 || index > this.size) {
      throw new RangeError("Index out of bounds");
    }

    let current = this.sentinel;
    for (let i = 0; i < index; i++) {
      current = current.next;
    }

    const newNode = new Node(value, current, current.next);
    current.next.prev = newNode;
    current.next = newNode;
    this.size++;
  }

  addFirst(value) {
    this.add(0, value);
  }

  addLast(value) {
    // TODO
    const newNode = new Node(value, this.sentinel.prev, this.sentinel);
    this.sentinel.prev.next = newNode;
    this.sentinel.prev = newNode;
    this.size++;
  }

  get(index) {
    if (index < 0 || index >= this.size) {
      throw new RangeError("Index out of bounds");
    }

    let current = this.sentinel.next;
    for (let i = 0; i < index; i++) {
      current = current.next;
    }

    return current.value;
  }

  getFirst() {
    return this.get(0);
  }

  getLast() {
    // TODO
    if (this.isEmpty()) {
    }

    return this.sentinel.prev.value;
  }

  remove(index) {
    if (index < 0 || index >= this.size) {
      throw new RangeError("Index out of bounds");
    }

    let current = this.sentinel.next;
    for (let i = 0; i < index; i++) {
      current = current.next;
    }

    current.prev.next = current.next;
    current.next.prev = current.prev;
    this.size--;

    return current.value;
  }

  removeFirst() {
    return this.remove(0);
  }

  removeLast() {
    // TODO
    const removedValue = this.sentinel.prev.value;
    this.sentinel.prev = this.sentinel.prev.prev;
    this.sentinel.prev.next = this.sentinel;
    this.size--;

    return removedValue;
  }

  size() {
    return this.size;
  }

  isEmpty() {
    return this.size === 0;
  }

  indexOf(value) {
    let current = this.sentinel.next;
    for (let i = 0; i < this.size; i++) {
      if (current.value === value) {
        return i;
      }
      current = current.next;
    }
    return -1;
  }

  contains(value) {
    return this.indexOf(value) !== -1;
  }

  toString() {
    let result = "[";
    let current = this.sentinel.next;
    while (current !== this.sentinel) {
      result += current.value;
      if (current.next !== this.sentinel) {
        result += ", ";
      }
      current = current.next;
    }
    result += "]";
    return result;
  }

  static arrayToDLL(arr) {
    const dll = new DLL();
    for (const value of arr) {
      dll.addLast(value);
    }
    return dll;
  }
}

// Example usage
const arr = [1, 2, 3, 4, 5];
const dll = DLL.arrayToDLL(arr);
console.log(dll.toString());

console.log(dll.size());
console.log(dll.isEmpty());

console.log(dll.indexOf(3));
console.log(dll.contains(3));
console.log(dll.indexOf(10));
console.log(dll.contains(10));

dll.add(2, 10);
console.log(dll.toString());

dll.addFirst(20);
console.log(dll.toString());

dll.addLast(30);
console.log(dll.toString());

console.log(dll.get(2));
console.log(dll.getFirst());
console.log(dll.getLast());

console.log(dll.remove(2));
console.log(dll.toString());
console.log(dll.removeFirst());
console.log(dll.toString());
console.log(dll.removeLast());
console.log(dll.toString());
